taichi/codegen/codegen_metal.cpp

```diff
@@ -634,36 +636,80 @@ class KernelCodegen : public IRVisitor {
     emit_mtl_kernel_func_sig(mtl_kernel_name, ka.buffers);
 
     const int sn_id = stmt->snode->id;
-    ka.num_threads = compiled_structs_->snode_descriptors.find(sn_id)
-                         ->second.total_num_elems_from_root;
+    // struct_for kernels use grid-stride loops
+    ka.num_threads = std::min(compiled_structs_->snode_descriptors.find(sn_id)
+                                  ->second.total_num_elems_from_root,
+                              kMaxNumThreadsGridStrideLoop);
+    if (!mtl_kernels_attribs_.empty() &&
+        mtl_kernels_attribs_.back().task_type ==
+            OffloadedStmt::TaskType::listgen) {
+      // TODO(#689): this is somewhat hacky. Once the IR is more polished, we
+      // can easily iterate through all the offloaded tasks beforehand to remove
+      // the (clear, listgen) for leaf node, instead of doing it here.
+      TI_ASSERT(mtl_kernels_attribs_.size() >= 2);
+      for (int i = 0; i < 2; ++i) {
+        TI_ASSERT(
+            mtl_kernels_attribs_.back().runtime_list_op_attribs.snode->id ==
+            sn_id);
+        mtl_kernels_attribs_.pop_back();
+      }
+    }
```

Comment:
I'm actually slightly confused by this part. By default the `offload` pass ignores the `listgen` for the leaf blocks - does the Metal backend introduce one more set of `clear` and `listgen` compared to the LLVM backend (so that we have to remove them here)?

Reply:
>  the offload pass ignores the listgen for the leaf blocks

Good point... I might have abused the terms a bit. For example:

```py
x = ti.var(ti.i32)
ti.root
  .dense(ti.i, (4,))  # S1
  .bitmasked(ti.i, (4,))  # S2
  .place(x)  # S3
```

I guess you were referring `S3` as the leaf node? In Metal's implementation, `S3` is just embedded inside `S2` as a struct member. So the actual leaf node is `S2`..

Reply:
Nice example.

```python
ti.init(print_ir=True)
x = ti.var(ti.i32)
ti.root.dense(ti.i, 4).bitmasked(ti.i, 4).place(x)

@ti.kernel
def loop():
    for i in x:
        print(x[i])

loop()
```

Generated IR on CUDA: (I modified `listgen S1dense` into `S0root->S1dense` to make things clearer)
```
kernel {
  $0 = offloaded clear_list S1dense
  $1 = offloaded listgen S0root->S1dense
  $2 = offloaded clear_list S2bitmasked
  $3 = offloaded listgen S1dense->S2bitmasked
  $4 = offloaded struct_for(S2bitmasked) block_dim=0 {
    <i32 x1> $5 = loop index 0
    <gen*x1> $6 = get root
    <i32 x1> $7 = const [0]
    <gen*x1> $8 = [S0root][root]::lookup($6, $7) activate = false
    <gen*x1> $9 = get child [S0root->S1dense] $8
    <i32 x1> $10 = bit_extract($5 + 0, 2~4)
    <gen*x1> $11 = [S1dense][dense]::lookup($9, $10) activate = false
    <gen*x1> $12 = get child [S1dense->S2bitmasked] $11
    <i32 x1> $13 = bit_extract($5 + 0, 0~2)
    <gen*x1> $14 = [S2bitmasked][bitmasked]::lookup($12, $13) activate = false
    <i32*x1> $15 = get child [S2bitmasked->S3place_i32] $14
    <i32 x1> $16 = global load $15
    print x[i], $16
  }
}
```

As you can see we ignore `listgen S2bitmasked->S3place`. I'm wondering which offloaded tasks are removed by the statements here?

Reply:
`S1dense->S2bitmasked` is removed.

```
kernel {
  $0 = offloaded clear_list S1dense
  $1 = offloaded listgen S0root->S1dense
  $4 = offloaded struct_for(S2bitmasked) block_dim=0 {
  ...
  }
}
```

Note that for `listgen` and `struct_for` on Metal, multiple threads could map to the same parent element in `ListManager`. Each thread covers the children of that parent element whose linearized index is within `[begin, begin+range)`. 

Concretely, after `$1`, `S1dense`'s `ListManager` will contain 4 elements. Then we have 16 threads for `struct_for(S2bitmasked)`, with `range = 1`. Every 4 threads will read the same element of `S1dense`, but they each handle a unique item in `S2bitmasked`.

Reply:
> Concretely, after `$1`, `S1dense`'s `ListManager` will contain 4 elements.

I see where divergence happens. In the LLVM impl, `$1` will generate one element, which is the only instance of `S1dense`.

Reply:
Yeah, now i also know what's wrong now. Sounds like Metal has an off-by-one problem when defining the SNode structs... 🙄